### PR TITLE
Updated sprimber engine and reporting enhancements 

### PR DESCRIPTION
**Pull request by vvolverine**

### CODE REVIEW

File: sprimber-engine/src/main/java/com/griddynamics/qa/sprimber/engine/Node.java [hunk 1]
```diff
@@ -0,0 +1,520 @@
...
+    private final String role;
+    private final Map<String, Object> attributes = new HashMap<>();
+
+    // core related properties
+    private Status status;
+    private Phase phase;
+    private Type type;
+    private boolean isBypassed;
+    /**
+     * The node can be in one the 3 state - SKIP, ERROR and normal execution.
```
Comment (ZaitsevaOlga) [inline sprimber-engine/src/main/java/com/griddynamics/qa/sprimber/engine/Node.java]: Looks like it's out of date. Please update.

File: sprimber-engine/src/main/java/com/griddynamics/qa/sprimber/engine/Node.java [hunk 2]
```diff
@@ -0,0 +1,520 @@
...
+    public void bypassExecution() {
+        this.isBypassed = true;
+    }
+
+    public boolean isBypassed() {
+        return this.isBypassed;
+    }
+
+    public boolean isEmptyHolder() {
+        return Type.HOLDER.equals(this.type) && this.children.size() == 0;
```
Comment (ZaitsevaOlga) [inline sprimber-engine/src/main/java/com/griddynamics/qa/sprimber/engine/Node.java]: Isn't it better to use some isEmpty() method?

File: sprimber-engine/src/main/java/com/griddynamics/qa/sprimber/engine/Node.java [hunk 3]
```diff
@@ -0,0 +1,520 @@
...
+     * * SCHEDULED - if the node marked for differed execution once the node picked up by executor
+     * and passed to the executor
+     * * EXECUTING - means that node currently in executing state
+     * * COMPLETED - execution if over, for details check the {@link Status}
+     */
+    enum Phase {
+        CREATED, PREPARED, SCHEDULED, EXECUTING, COMPLETED
+    }
+
+    enum Type {
```
Comment (ZaitsevaOlga) [inline sprimber-engine/src/main/java/com/griddynamics/qa/sprimber/engine/Node.java]: Don't we need javadocs here also?

File: sprimber-engine/src/main/java/com/griddynamics/qa/sprimber/reporting/StepExecutionReportCatcher.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+//    @Autowired
+//    public void setStepDefinitionsFactory(SpringTestMethodsLoader stepDefinitionsFactory) {
+//        StepExecutionReportCatcher.stepDefinitionsFactory = stepDefinitionsFactory;
+//    }
+
+    @Around("stepMethodsExecution()")
+    public Object surroundStepExecution(ProceedingJoinPoint joinPoint) throws Throwable {
+        Object result = null;
+        result = joinPoint.proceed();
+//        ExecutionResult executionResult = new ExecutionResult(PASSED);
```
Comment (ZaitsevaOlga) [inline sprimber-engine/src/main/java/com/griddynamics/qa/sprimber/reporting/StepExecutionReportCatcher.java]: Why is it commented?

File: sprimber-engine/src/main/java/com/griddynamics/qa/sprimber/stepdefinition/TestMethodRegistry.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    public Stream<TestMethod> streamAllTestMethods() {
+        return testMethodMap.values().stream();
+    }
+
+    public Stream<TestMethod> streamByStyle(String style) {
+        return testMethodMap.values().stream()
+                .filter(testMethod -> style.equals(testMethod.getStyle()));
+    }
+
+//    public Stream<Node> provideBeforeTestHookNodes() {
```
Comment (ZaitsevaOlga) [inline sprimber-engine/src/main/java/com/griddynamics/qa/sprimber/stepdefinition/TestMethodRegistry.java]: Why is it commented?

File: pom.xml [hunk 6]
```diff
@@ -111,20 +114,20 @@
...
-                    <execution>
-                        <id>sign-artifacts</id>
-                        <phase>verify</phase>
-                        <goals>
-                            <goal>sign</goal>
-                        </goals>
-                    </execution>
-                </executions>
-            </plugin>
+<!--            <plugin>-->
```
[the remark this answers is not in the document]
Comment (vvolverine) [reply pom.xml]: Moved under the profile and uncommented

File: sprimber-classic/src/test/java/com/griddynamics/qa/sprimber/discovery/ClassicTestBinderTest.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+/**
+ * @author fparamonov
+ */
+public class ClassicTestBinderTest {
+
+    private static final String DEFAULT_HISTORY_ID = "ee17157fc5eba979689e028b6132883d";
+
+    @Test
+    public void testSimpleBinding() {
+//        ClassicStepFactory stepFactoryMock = Mockito.mock(ClassicStepFactory.class);
```
[the remark this answers is not in the document]
Comment (vvolverine) [reply sprimber-classic/src/test/java/com/griddynamics/qa/sprimber/discovery/ClassicTestBinderTest.java]: need refactoring not critical for now.

File: sprimber-engine/src/main/java/com/griddynamics/qa/sprimber/engine/Node.java (hunk 2 above)
Comment (vvolverine) [reply sprimber-engine/src/main/java/com/griddynamics/qa/sprimber/engine/Node.java]: updated with suggestion

File: sprimber-engine/src/main/java/com/griddynamics/qa/sprimber/engine/Node.java (hunk 3 above)
Comment (vvolverine) [reply sprimber-engine/src/main/java/com/griddynamics/qa/sprimber/engine/Node.java]: added

File: sprimber-engine/src/main/java/com/griddynamics/qa/sprimber/reporting/StepExecutionReportCatcher.java (hunk 4 above)
Comment (vvolverine) [reply sprimber-engine/src/main/java/com/griddynamics/qa/sprimber/reporting/StepExecutionReportCatcher.java]: need refactoring for future, currently this feature not implemented and not included into final configuration

File: sprimber-engine/src/main/java/com/griddynamics/qa/sprimber/stepdefinition/TestMethodRegistry.java (hunk 5 above)
Comment (vvolverine) [reply sprimber-engine/src/main/java/com/griddynamics/qa/sprimber/stepdefinition/TestMethodRegistry.java]: unused code removed

File: sprimber-cucumber/src/main/java/com/griddynamics/qa/sprimber/discovery/PickleStepFactory.java [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+                .filter(testMethod -> StringUtils.isNotBlank(testMethod.getTextPattern()))
+                .filter(filterTestMethodByTags())
+                .map(testMethod ->
+                        stepMatcher.matchAndGetArgumentsFrom(stepCandidate, testMethod, testMethod.getMethod().getParameterTypes())
+                                .map(arguments -> addStepContainerNode(stepContainerNode, stepCandidate, testMethod, arguments))
+                )
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .collect(Collectors.toList());
+        if (nodeList.size() == 0) {
```
Comment (ZaitsevaOlga) [inline sprimber-cucumber/src/main/java/com/griddynamics/qa/sprimber/discovery/PickleStepFactory.java]: Maybe it would be better to replace to isEmpty?

**Pull request merged**